packages/ttypescript/src/loadTypescript.ts

```diff
@@ -1,32 +1,66 @@
-import * as fs from 'fs';
-import * as resolve from 'resolve';
 import * as TS from 'typescript';
 type ts = typeof TS;
+import { readFileSync } from 'fs';
+import { sync as resolveSync } from 'resolve';
 import { patchCreateProgram } from './patchCreateProgram';
 import { dirname } from 'path';
 import { runInThisContext } from 'vm'
+import Module = require('module')
 
+type TypeScriptModuleFactory = (exports: ts, require: NodeRequireFunction, module: { exports: any }, filename: string, dirname: string) => ts;
+const typeScriptModuleFactoryCache: { [key: string]: TypeScriptModuleFactory } = Object.create(null)
 
 export function loadTypeScript(
     filename: string,
     { folder = __dirname, forceConfigLoad = false }: { folder?: string; forceConfigLoad?: boolean } = {}
 ): ts {
-    const opts = { basedir: folder };
-    const typescriptFilename = resolve.sync('typescript/lib/' + filename, opts);
-
-    const m = { exports: {} as ts }
-    const code = fs.readFileSync(typescriptFilename, 'utf8');
-    runInThisContext(
-        `(function (exports, require, module, __filename, __dirname) {${code}\n});`,
-        { filename: typescriptFilename, lineOffset: 0, displayErrors: true }
-    ).call(m.exports, m.exports, require, m, typescriptFilename, dirname(typescriptFilename));
-    
+    const libFilename = resolveSync('typescript/lib/' + filename, { basedir: folder })
+
+    let factory = typeScriptModuleFactoryCache[libFilename];
+    if (!factory) {
+        const code = readFileSync(libFilename, 'utf8');
+        factory = runInThisContext(
+            `(function (exports, require, module, __filename, __dirname) {${code}\n});`,
+            { filename: libFilename, lineOffset: 0, displayErrors: true }
+        )
+        typeScriptModuleFactoryCache[libFilename] = factory;
+    }
+
+    if (!require.cache[libFilename]) {
+        require.cache[libFilename] = createPristineTypeScriptLazyModule(libFilename, factory);
+    }
+
+    const m = { exports: {} as ts };
+    factory.call(m.exports, m.exports, require, m, libFilename, dirname(libFilename));
     const ts = m.exports;
 
     const [major, minor] = ts.versionMajorMinor.split('.');
     if (+major < 3 && +minor < 7) {
         throw new Error('ttypescript supports typescript from 2.7 version');
     }
-    patchCreateProgram(ts, forceConfigLoad);
-    return ts;
+
+    return patchCreateProgram(ts, forceConfigLoad);
 }
+
+function createPristineTypeScriptLazyModule(filename: string, factory: TypeScriptModuleFactory) {
+    const m = new Module(filename, module);
+    m.filename = filename;
+    m.loaded = true;
+    m.paths = module.paths.slice();
+    let ts: ts | undefined
+    Object.defineProperty(m, 'exports', {
+        get() {
+            if (ts === undefined) {
+                ts = {} as ts
+                factory.call(ts, ts, require, m, filename, dirname(filename));
```

Comment:
I think it is incorrect, because if you do `module.exports = something`,  `exports` variable will be empty
```
factory.call(m.exports, m.exports, require, m, filename, dirname(filename));
ts = m.exports;
```

Reply:
It works because the loaded code will do module.exports = ts. But I can make it more clear